# Load data

```{r}
# metadata file
source('scripts/Utilities.R', chdir = TRUE)
colony_metadata <- read.csv("data/colony_metadata.tsv", sep = "\t", header = T)
metadata_df <- read.csv("data/sample_extraction_metadata.tsv", sep = "\t", header = T) %>%
                    mutate(Sample = Vectorize(sample_name_from_shortname)(Label)) %>%
                    left_join(colony_metadata, by=c("Colony"="Colony_ID"))
colnames(metadata_df)
```

# Dataset summary

```{r}
table(metadata_df$Colony_type)
```
 M  SM   W 
 15  15 130 

```{r}
table(metadata_df$Species)
```
A. m. carnica   A. m. jemenica   A. m. lamarkii   A. m. littorea 
              10               25               25               25 
 A. m. monticola A. m. scutellata   A. m. simensis 
              25               25               25

```{r}
table(metadata_df$Species, metadata_df$Colony_type)
```
                    M SM  W
  A. m. carnica    10  0  0
  A. m. jemenica    0  0 25
  A. m. lamarkii    0  0 25
  A. m. littorea    5 15  5
  A. m. monticola   0  0 25
  A. m. scutellata  0  0 25
  A. m. simensis    0  0 25

```{r}
table(metadata_df$Species, metadata_df$Location_name)
```

```{r}
table(metadata_df$Location_name)
```
Aberdares      Apiary Hohenheim          Arabuko edge 
                    5                    10                     5 
               Entaye                  Gede         Karura forest 
                    5                    10                    10 
               Kilifi            Mau Forest    Mount Kenya Forest 
                    5                     5                     5 
             Mt Kenya                Mtwapa      Muranga, Samburu 
                    5                     5                     5 
             Nguruman Nyambene Hills Forest                 Sekem 
                   10                     5                    25 
              Turkana            Wendogenet                 Werie 
                   25                    10                    10

# Raw data summary and QC result

Paired reads were generated acros single or multiple lanes for each sample.

Generate tsv files `number_of_reads.tsv` and `number_of_reads_mapping.tsv` for visualization using `scripts/Check_seq_lengths.py`.

## Library depth

### Across dataset

```{r}
df_numbers <- read.csv("results/visualization/number_of_reads.tsv", sep="\t", row.names=NULL, header=T)
ggplot() +
    geom_bar(data=df_numbers, aes(x=sample, y=raw_reads), stat="identity") +
    labs(title="Number of reads per sample", x="Sample", y="Number of reads") +
    scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
    make_theme(x_angle=90, x_size=5, x_hj=1, x_vj=0.5)
ggsave("results/figures/00-total_number_of_reads.pdf")
```

## Trimming results

### Trimming result across dataset

### Trimming result per sample

## Host filtering results

```{r}
df_mapping_results <- read.csv("results/visualization/number_of_reads_mapping.tsv", sep="\t", row.names=NULL, header=T) %>%
                        left_join(metadata_df %>% select(Sample, Species, Location_name, Colony_type), by=c("sample"="Sample"))
df_plot <- df_mapping_results %>%
            select(sample, total_reads, host_reads, Species) %>%
            mutate(non_host_mapped = total_reads - host_reads) %>%
            select(-total_reads) %>%
            gather(key="type", value="reads", -sample, -Species)
ggplot() +
    geom_bar(data=df_plot, aes(x=sample, y=reads, fill=type), stat="identity") +
    labs(title="Number of host reads per sample", x="Sample", y="Number of host reads") +
    scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
    facet_wrap(~Species, scales = "free_x") +
    make_theme(x_angle=90, x_size=5, x_hj=1, x_vj=0.5)
ggsave("results/figures/01-number_of_host_reads_species.pdf")
```

```{r}
df_plot <- df_mapping_results %>%
            select(sample, total_reads, host_mapped_depletion, Location_name) %>%
            mutate(reads_for_assembly = total_reads - host_mapped_depletion) %>%
            select(-total_reads) %>%
            gather(key="type", value="reads", -sample, -Location_name)
ggplot() +
    geom_bar(data=df_plot, aes(x=sample, y=reads, fill=type), stat="identity") +
    labs(title="Number of host reads per sample", x="Sample", y="Number of host reads") +
    scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
    facet_wrap(~Location_name, scales = "free_x") +
    make_theme(x_angle = 45, x_size = 7, x_hj = 1, x_vj = 1)
ggsave("results/figures/01-number_of_host_reads_location.pdf")
```

# Assembly results

## Assembly results - read from quast

```{r}
quast_df <- data.frame()
for (sample in metadata_df$Sample) {
    quast_file <- paste("results/03_Assembly/", sample, "/", sample, "_quast_report/transposed_report.tsv", sep="")
    if (file.exists(quast_file)) {
        quast_df <- rbind(quast_df, read.csv(quast_file, sep="\t", header=T))
    }
}
quast_df_filt <- quast_df %>%
                    filter(grepl("filtered", Assembly)) %>%
                    mutate(Assembly = gsub("_scaffolds_1k_filtered", "", Assembly)) %>%
                    left_join(metadata_df %>% select(Sample, Species, Location_name, Colony_type), by=c("Assembly"="Sample"))
```
### Total length of assembly

```{r}
ggplot() +
    geom_bar(data=quast_df_filt, aes(x=Assembly, y=Total.length), stat="identity") +
    labs(title="Total length of assembly", x="Assembly", y="Total length") +
    scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
    facet_wrap(~Species, scales = "free_x") +
    make_theme(x_angle=90, x_size=5, x_hj=1, x_vj=0.5)
ggsave("results/figures/03-total_length_of_assembly_species.pdf")
```

### Contigs bigger than or equal to 10000 bp

```{r}
ggplot() +
    geom_bar(data=quast_df_filt, aes(x=Assembly, y=X..contigs.....10000.bp.), stat="identity") +
    labs(title="Contigs >= 10000bp of assembly", x="Assembly", y="Contigs >= 10000bp") +
    facet_wrap(~Species, scales = "free_x") +
    make_theme(x_angle=90, x_size=5, x_hj=1, x_vj=0.5)
ggsave("results/figures/03-Contigs_geq_10000bp_of_assembly_species.pdf")
```

### Total length bigger than or equal to 10000 bp

```{r}
ggplot() +
    geom_bar(data=quast_df_filt, aes(x=Assembly, y=Total.length.....1000.bp.), stat="identity") +
    labs(title="Total length >= 10000bp of assembly", x="Assembly", y="Total length >= 10000bp") +
    scale_y_continuous(labels=unit_format(unit = "M", scale = 1e-6)) +
    facet_wrap(~Species, scales = "free_x") +
    make_theme(x_angle=90, x_size=5, x_hj=1, x_vj=0.5)
ggsave("results/figures/03-Total_length_geq_10000bp_of_assembly_species.pdf")
```

# Simka k-mer similarity

```{r}
matabjac <- as.matrix(read.csv("results/04_binning_backmapping/simka_out/mat_abundance_jaccard.csv.gz", sep = ";", header = T, row.names = 1))
matabjac <- (1 - matabjac) * 100
# add color bar showing subspecies
Species_names <- matabjac %>% rownames() %>% as.data.frame() %>% left_join(metadata_df %>% select(Sample, Species), by=c("."="Sample")) %>% select(Species) %>% pull
top_annotation <- HeatmapAnnotation(Subspecies = Species_names,
                                    col = list(Subspecies = Species_colors),
                                    annotation_name_side = "left")
left_annotation <- rowAnnotation(Subspecies = Species_names,
                                    col = list(Subspecies = Species_colors),
                                    annotation_name_side = "bottom")
# make a complexHeatmap object
heatmap <- Heatmap(matabjac,
                   name = "Jaccard dissimilarity by k-mers",
                   col = colorRamp2(c(0, 10, 20, 40, 100), c("#f1eef6", "#bdc9e1", "#0570b0", "#023858", "#000000")),
                   cluster_rows = TRUE,
                   cluster_columns = TRUE,
                   top_annotation = top_annotation,
                   left_annotation = left_annotation,
                   show_row_names = TRUE,
                   show_column_names = TRUE)
pdf("results/figures/04-simka_kmer_similarity.pdf", width = 15, height = 15)
draw(heatmap, merge_legend = TRUE)
dev.off()

# pcoa
pcoa <- cmdscale(as.dist(matabjac), k=2)
pcoa_df <- data.frame(Sample = rownames(matabjac), PC1 = pcoa[,1], PC2 = pcoa[,2]) %>%
            left_join(metadata_df %>% select(Sample, Species, Location_name, Colony_type), by=c("Sample"="Sample"))
ggplot() +
    geom_point(data=pcoa_df, aes(x=PC1, y=PC2, color=Species, size = 5)) +
    labs(title="PCoA of k-mer similarity", x="PC1", y="PC2") +
    xlim(-3,3) +
    ylim(-3,3) +
    scale_color_manual(values = Species_colors) +
    make_theme(setCol = F)
ggsave("results/figures/04-simka_kmer_similarity_pcoa.pdf")

```